pkg/l2announcer/l2announcer.go

```diff
@@ -0,0 +1,1073 @@
+	if ns == "" {
+		ns = metav1.NamespaceDefault
+	}
+
+	return ns
+}
+
+const leasePrefix = "cilium-l2announce"
+
+func (l2a *L2Announcer) newLeaseLock(svc *slim_corev1.Service) *resourcelock.LeaseLock {
+	return &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Namespace: l2a.leaseNamespace(),
+			Name:      fmt.Sprintf("%s-%s-%s", leasePrefix, svc.Namespace, svc.Name),
+		},
+		Client: l2a.Params.Clientset.CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: l2a.localNode.Name,
+		},
+	}
+}
+
+// Check all selected services, delete services which are no longer selected by any of the policies.
+func (l2a *L2Announcer) gcOrphanedServices() error {
+	for _, ss := range l2a.selectedServices {
+		err := l2a.gcOrphanedService(ss)
+		if err != nil {
+			return fmt.Errorf("gcOrphanedService: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (l2a *L2Announcer) gcOrphanedService(ss *selectedService) error {
+	// Only GC policies that have been orphaned (all policies that created it has gone away)
+	if len(ss.byPolicies) > 0 {
+		return nil
+	}
+
+	// Stop leader election routine
+	ss.stop()
+
+	// Recalculation will remove all entries since we stopped the leader election.
+	if err := l2a.recalculateL2EntriesTableEntries(ss); err != nil {
+		return fmt.Errorf("recalculateNeighborProxyTableEntries: %w", err)
+	}
+
+	// Remove service from selected services
+	delete(l2a.selectedServices, serviceKey(ss.svc))
+	return nil
+}
+
+func (l2a *L2Announcer) processLocalNodeEvent(ctx context.Context, event resource.Event[*v2.CiliumNode]) error {
+	var err error
+	if event.Kind == resource.Upsert {
+		err = l2a.upsertLocalNode(ctx, event.Object)
+		if err != nil {
+			err = fmt.Errorf("upsert local node: %w", err)
+		}
+	}
+
+	event.Done(err)
+	return err
+}
+
+func (l2a *L2Announcer) upsertLocalNode(ctx context.Context, localNode *v2.CiliumNode) error {
+	// If the label set did not change, nothing to do.
+	if l2a.localNode != nil && labels.Equals(l2a.localNode.Labels, labels.Set(localNode.Labels)) {
+		return nil
+	}
+
+	l2a.localNode = localNode
+
+	// Delete any policies that no longer match the new label set
+	var errs error
+	for key, selectedPolicy := range l2a.selectedPolicies {
+		var nodeselector labels.Selector
+		if selectedPolicy.policy.Spec.NodeSelector == nil {
+			nodeselector = labels.Everything()
+		} else {
+			var err error
+			nodeselector, err = slim_meta_v1.LabelSelectorAsSelector(selectedPolicy.policy.Spec.NodeSelector)
+			if err != nil {
+				if err2 := l2a.updatePolicyStatus(ctx, selectedPolicy.policy, "io.cilium/bad-node-selector", err); err2 != nil {
+					l2a.Params.Logger.WithError(err2).Error("updating policy status failed")
+				}
+				return fmt.Errorf("make node selector: %w", err)
+			}
+		}
+		if err := l2a.updatePolicyStatus(ctx, selectedPolicy.policy, "io.cilium/bad-node-selector", nil); err != nil {
+			l2a.Params.Logger.WithError(err).Error("updating policy status failed")
```

Comment:
Yea. This call is meant to correct the status of the policy in case the status indicated the policy was invalid before. If that goes wrong then the status might not be up-to-date, but that should not keep us from changing our desired state.